staging/src/k8s.io/apiserver/plugin/pkg/audit/dynamic/dynamic.go

```diff
@@ -0,0 +1,295 @@
+	delegate      audit.Backend
+	stopChan      chan struct{}
+}
+
+// NewBackend returns a backend that dynamically updates its configuration
+// based on a shared informer.
+func NewBackend(c *Config) (audit.Backend, error) {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventSink := &v1core.EventSinkImpl{Interface: c.KubeClient.CoreV1().Events("")}
+	eventBroadcaster.StartRecordingToSink(eventSink)
+
+	scheme := runtime.NewScheme()
+	err := auditregv1alpha1.AddToScheme(scheme)
+	if err != nil {
+		return nil, err
+	}
+	recorder := eventBroadcaster.NewRecorder(scheme, v1.EventSource{Component: "apiserver"})
+
+	if c.BufferedConfig == nil {
+		defaultConfig := NewDefaultWebhookBatchConfig()
+		c.BufferedConfig = &defaultConfig
+	}
+	cm, err := webhook.NewClientManager(auditv1.SchemeGroupVersion, func(s *runtime.Scheme) error {
+		auditinstall.Install(s)
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	authInfoResolver, err := webhook.NewDefaultAuthenticationInfoResolver("")
+	if err != nil {
+		return nil, err
+	}
+	cm.SetAuthenticationInfoResolver(authInfoResolver)
+	cm.SetServiceResolver(webhook.NewDefaultServiceResolver())
+	cm.SetAuthenticationInfoResolverWrapper(c.AuthInfoResolverWrapper)
+
+	manager := &dynamic{
+		config:               c,
+		backends:             &atomic.Value{},
+		backendUpdateMutex:   sync.Mutex{},
+		webhookClientManager: cm,
+		recorder:             recorder,
+	}
+	manager.backends.Store(syncedBackends{})
+
+	c.Informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			manager.addSink(obj.(*auditregv1alpha1.AuditSink))
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			manager.updateSink(oldObj.(*auditregv1alpha1.AuditSink), newObj.(*auditregv1alpha1.AuditSink))
+		},
+		DeleteFunc: func(obj interface{}) {
+			manager.deleteSink(obj.(*auditregv1alpha1.AuditSink))
+		},
+	})
+
+	return manager, nil
+}
+
+type dynamic struct {
+	// backendUpdateMutex holds an update lock on the backends
+	backendUpdateMutex   sync.Mutex
+	config               *Config
+	backends             *atomic.Value
+	webhookClientManager webhook.ClientManager
+	recorder             record.EventRecorder
+}
+
+type syncedBackends map[types.UID]*Backend
+
+// Names returns the names of the backend configurations
+func (s syncedBackends) Names() []string {
+	names := []string{}
+	for _, backend := range s {
+		names = append(names, backend.configuration.Name)
+	}
+	return names
+}
+
+// ProcessEvents proccesses the given events per current backend map
+func (d *dynamic) ProcessEvents(events ...*auditinternal.Event) {
+	for _, b := range d.GetBackends() {
+		b.delegate.ProcessEvents(events...)
+	}
+}
+
+// Run the dynamic backends
+func (d *dynamic) Run(stopCh <-chan struct{}) error {
+	var funcs []func() error
+	for _, b := range d.GetBackends() {
+		funcs = append(funcs, func() error {
+			return b.delegate.Run(b.stopChan)
+		})
+	}
```

Comment:
Please ensure these cases are covered in the unit test.

Reply:
I still think we should postpone adding the handler until Run is called. I could see some weird race condition being triggered if a dynamic backend is run before Run is called on the whole audit pipeline...

Reply:
So I solved this by adding a `started` field to the struct, and a new sink will fail to provision and send an event informing that the backend hasn't been ran. This way we get an explicit message returned about the error. I'm not sure what would happen if a sink were added without the event handler in place, if we want to go that way I would just need to sync with an informer specialist.

Reply:
Got it, please add a comment explaining this. For my own edification, what starts the informer?

Reply:
sounds good, the informers are started with a post start hook [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/config.go#L494), considering the audit backends are ran [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L287) it would in theory be possible for that `Run` method to hang, and the informers to start before it finishes. Worst case scenario is the informer pauses right on boot to [resync state](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L330) which would block any other processes using that informer. This seems reasonable though with the implementation, the only way I can see to totally avoid this would be to call the `Run` method serially on the backends before boot.

Reply:
I suppose if we are really concerned about that another option would be to create a separate informer outside of the shared informer factory, and run it ourselves. This would get rid of the slight possibility of a resync pause for a second listener. This comes with a bit of overhead though and it feels like we would be solving for problems we don't have.